logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/ConversationVerificationStatusHandler.kt

```diff
@@ -0,0 +1,84 @@
+import com.wire.kalium.logic.data.conversation.ConversationRepository
+import com.wire.kalium.logic.data.conversation.ConversationVerificationStatus
+import com.wire.kalium.logic.data.conversation.ProtocolInfoMapper
+import com.wire.kalium.logic.data.message.Message
+import com.wire.kalium.logic.data.message.MessageContent
+import com.wire.kalium.logic.data.message.PersistMessageUseCase
+import com.wire.kalium.logic.data.user.UserId
+import com.wire.kalium.logic.di.MapperProvider
+import com.wire.kalium.logic.functional.flatMap
+import com.wire.kalium.logic.functional.getOrElse
+import com.wire.kalium.util.DateTimeUtil
+import com.wire.kalium.util.KaliumDispatcher
+import com.wire.kalium.util.KaliumDispatcherImpl
+import kotlinx.coroutines.withContext
+
+/**
+ * Notify user (by adding System message in conversation) if needed about changes of Conversation Verification status.
+ */
+internal interface ConversationVerificationStatusHandler {
+    suspend operator fun invoke(conversation: Conversation, status: ConversationVerificationStatus)
+}
+
+internal class ConversationVerificationStatusHandlerImpl(
+    private val conversationRepository: ConversationRepository,
+    private val persistMessage: PersistMessageUseCase,
+    private val selfUserId: UserId,
+    private val protocolInfoMapper: ProtocolInfoMapper = MapperProvider.protocolInfoMapper(),
+    kaliumDispatcher: KaliumDispatcher = KaliumDispatcherImpl
+) : ConversationVerificationStatusHandler {
+    private val dispatcher = kaliumDispatcher.io
+
+    override suspend fun invoke(conversation: Conversation, status: ConversationVerificationStatus): Unit = withContext(dispatcher) {
+        if (shouldNotifyUser(conversation, status)) {
+            val verificationDegradedMessage = Message.System(
+                id = uuid4().toString(),
+                content = MessageContent.VerificationDegraded(protocolInfoMapper.fromInfoToProtocol(conversation.protocol)),
+                conversationId = conversation.id,
+                date = DateTimeUtil.currentIsoDateTimeString(),
+                senderUserId = selfUserId,
+                status = Message.Status.SENT,
+                visibility = Message.Visibility.VISIBLE,
+                expirationData = null
+            )
+
+            persistMessage(verificationDegradedMessage)
+                .flatMap { conversationRepository.setInformedAboutDegradedMLSVerificationFlag(conversation.id, true) }
+        } else {
+            conversationRepository.setInformedAboutDegradedMLSVerificationFlag(conversation.id, false)
+        }
+    }
+
+    private suspend fun shouldNotifyUser(conversation: Conversation, status: ConversationVerificationStatus): Boolean =
+        if (status == ConversationVerificationStatus.DEGRADED) {
+            if (conversation.protocol is Conversation.ProtocolInfo.MLS) {
+                !conversationRepository.isInformedAboutDegradedMLSVerification(conversation.id).getOrElse(true)
```

Comment:
There is an edge case here, let's assume we have a conversation that was degraded, we go offline, it gets verified. once we get online again; before that it gets degraded again. should we notify the user again? I don't know tbh. we need to talk about this super rare edge case.

Reply:
i think we can keep as it is.
1. it's a rare case
2. nothing changed for user: conversation was degraded last time we opens it, and it's degraded now, so what the point of spamming user with system messages
3. there is no way to cover such a case, as we are getting verification status only on opening the conversation, there is no any web-socket event for it. So no way to find out how much time status changed since the last checking.